### PR TITLE
Got pytest cov to 80% - also added README install instructions

### PR DESCRIPTION
**Pull request by zandersnow14**
- Added a variety of tests - got the coverage up to 80%
- Added a conftest.py file to mock the api client
- Added a line in the README to install type hints - removed my_py_boto3 from requirements

### CODE REVIEW

File: pipeline/app.py [hunk 1]
```diff
@@ -219,7 +219,7 @@ def unsubscribe_index():
             return render_template('/unsubscribe/not_subscribed.html')
 
         user_products = get_products_from_email(conn, email)
-        print(user_products)
+
```
Comment (tadawson23) [inline pipeline/app.py:222]: Thanks for taking out the print statements

File: pipeline/conftest.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+import pytest
+from flask.testing import FlaskClient
+
+from app import app
+
+
+@pytest.fixture()
+def api_client() -> FlaskClient:
+    """Returns a version of the API for testing."""
```
Comment (tadawson23) [inline pipeline/conftest.py]: The docstring formatting needs to change :)

File: pipeline/test_app.py [hunk 3]
```diff
@@ -4,26 +4,254 @@
...
+    mock_execute = mock_db_connection.cursor().execute
+    mock_fetchone = mock_db_connection.cursor().fetchone
+
+    mock_fetchone.return_value = mock_data
+
+    insert_subscription_data(mock_db_connection, test_email, test_url)
+
+    assert mock_execute.call_count == 3
+
+
```
Comment (tadawson23) [inline pipeline/test_app.py:128]: Can remove commented out code?

File: pipeline/README.md [hunk 4]
```diff
@@ -8,7 +8,8 @@ The files in this folder are used scrape information from various websites, extr
 - It is recommended before stating any installations that you make a new virtual environment. 
 - A new environment will be required for each folder in this repository.
 
-- Install all requirements for this folder: `pip3 install -r requirements.txt`.
+- Install all requirements for this folder: `pip3 install -r requirements.txt`
+- Install type hints for boto3: `python -m pip install 'boto3-stubs[ses]'`
```
Comment (Abbey-Wilkinson) [inline pipeline/README.md:12]: Good remembrance.

File: pipeline/app.py [hunk 5]
```diff
@@ -102,11 +102,11 @@ def insert_subscription_data(conn: connection, user_email: str, product_url: str
 
     user_query = "SELECT user_id FROM users WHERE email = (%s);"
     cur.execute(user_query, (user_email,))
-    user_id = cur.fetchone()['user_id']
+    user_id = cur.fetchone().get('user_id')
 
     product_query = "SELECT product_id FROM products WHERE product_url = (%s);"
     cur.execute(product_query, (product_url,))
-    product_id = cur.fetchone()['product_id']
+    product_id = cur.fetchone().get('product_id')
```
Comment (Abbey-Wilkinson) [inline pipeline/app.py:109]: Good checking.

File: pipeline/conftest.py [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+from flask.testing import FlaskClient
+
+from app import app
+
+
+@pytest.fixture()
+def api_client() -> FlaskClient:
+    """
+    Returns a version of the API for testing.
+    """
```
Comment (Abbey-Wilkinson) [inline pipeline/conftest.py:15]: Thanks for adding the docstring.

File: pipeline/test_app.py [hunk 7]
```diff
@@ -2,28 +2,231 @@
...
+def test_submit_get(mock_render_template, mock_get_database_connection, mock_scrape_asos_page, mock_insert_user_data, mock_insert_product_data, mock_insert_subscription_data, api_client):
+    """
+    Tests a get request to the addproducts returns a 200 status code.
+    """
+
+    load_dotenv()
+
+    response = api_client.get("/addproducts")
+
+    assert response.status_code == 200
```
Comment (Abbey-Wilkinson) [inline pipeline/test_app.py:182]: Very big but necessary!

**Review by Abbey-Wilkinson — APPROVED** [review]

**Pull request merged**